File: benchmarks/drivers/typescript/bench_native.ts
/**
 * Benchmark MongoDB native Node.js driver for comparison against MongoCore.
 */

import { MongoClient, ObjectId } from 'mongodb';
import { readFileSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { performance } from 'perf_hooks';
import * as os from 'os';

// Load config
const CONFIG = JSON.parse(readFileSync(join(__dirname, '..', 'common.json'), 'utf-8'));
const DATA_DIR = join(__dirname, '..', '..', 'data');
const RESULTS_DIR = join(__dirname, '..', '..', 'results');
mkdirSync(RESULTS_DIR, { recursive: true });

const QUICK_MODE = process.argv.includes('--quick');

const WARMUP = QUICK_MODE ? 0 : CONFIG.warmup_iterations.typescript;
const MIN_TIME = QUICK_MODE ? 0 : CONFIG.min_time_secs;
const MAX_ITERS = QUICK_MODE ? 1 : CONFIG.max_iterations;
const MAX_TIME = QUICK_MODE ? 5 : CONFIG.max_time_secs;
const DB_NAME = CONFIG.database;

interface SystemInfo {
  os: string;
  arch: string;
  cpus: number;
  ram_gb: number;
  mongocore_version: string;
  driver: string;
}

interface BenchResult {
  benchmark: string;
  category: string;
  driver: string;
  dataset_size_bytes: number;
  batch_size: number;
  iterations: number;
  total_time_secs: number;
  ops_per_sec: number;
  mb_per_sec: number;
  percentiles: {
    p10: number;
    p25: number;
    p50: number;
    p75: number;
    p90: number;
    p95: number;
    p99: number;
  };
  timestamp: string;
  system: SystemInfo;
}

function getSystemInfo(): SystemInfo {
  const totalMem = os.totalmem();
  return {
    os: os.platform(),
    arch: os.arch(),
    cpus: os.cpus().length,
    ram_gb: Math.round((totalMem / (1024 ** 3)) * 10) / 10,
    mongocore_version: 'native',
    driver: 'mongodb-node',
  };
}

function percentile(data: number[], pct: number): number {
  const idx = Math.max(0, Math.ceil(data.length * pct / 100) - 1);
  return data[Math.min(idx, data.length - 1)];
}

async function runBenchmark(
  name: string,
  category: string,
  setupFn: (client: MongoClient) => Promise<void>,
  beforeTaskFn: (client: MongoClient) => Promise<void>,
  taskFn: (client: MongoClient) => Promise<void>,
  afterTaskFn: (client: MongoClient) => Promise<void>,
  teardownFn: (client: MongoClient) => Promise<void>,
  datasetSizeBytes: number,
  batchSize: number,
): Promise<BenchResult> {
  const client = new MongoClient(CONFIG.mongodb_uri);
  await client.connect();

  await setupFn(client);

  // Warmup
  for (let i = 0; i < WARMUP; i++) {
    await beforeTaskFn(client);
    await taskFn(client);
    await afterTaskFn(client);
  }

  // Timed iterations
  const times: number[] = [];
  let totalTime = 0.0;
  let iteration = 0;

  while (totalTime < MIN_TIME || iteration < 5) {
    if (iteration >= MAX_ITERS || totalTime >= MAX_TIME) {
      break;
    }

    await beforeTaskFn(client);
    const start = performance.now();
    await taskFn(client);
    const elapsed = (performance.now() - start) / 1000;
    await afterTaskFn(client);

    times.push(elapsed);
    totalTime += elapsed;
    iteration++;
  }

  await teardownFn(client);
  await client.close();

  // Calculate metrics
  times.sort((a, b) => a - b);
  const median = times[Math.floor(times.length / 2)];
  const opsPerSec = batchSize / median;
  const mbPerSec = datasetSizeBytes / median / 1_000_000;

  const result: BenchResult = {
    benchmark: name,
    category: category,
    driver: 'mongodb-node',
    dataset_size_bytes: datasetSizeBytes,
    batch_size: batchSize,
    iterations: times.length,
    total_time_secs: Math.round(totalTime * 1000) / 1000,
    ops_per_sec: Math.round(opsPerSec * 10) / 10,
    mb_per_sec: Math.round(mbPerSec * 1000) / 1000,
    percentiles: {
      p10: Math.round(percentile(times, 10) * 1_000_000) / 1_000_000,
      p25: Math.round(percentile(times, 25) * 1_000_000) / 1_000_000,
      p50: Math.round(median * 1_000_000) / 1_000_000,
      p75: Math.round(percentile(times, 75) * 1_000_000) / 1_000_000,
      p90: Math.round(percentile(times, 90) * 1_000_000) / 1_000_000,
      p95: Math.round(percentile(times, 95) * 1_000_000) / 1_000_000,
      p99: Math.round(percentile(times, 99) * 1_000_000) / 1_000_000,
    },
    timestamp: new Date().toISOString(),
    system: getSystemInfo(),
  };

  console.log(`  ${name}: ${opsPerSec.toFixed(0)} ops/s, ${mbPerSec.toFixed(2)} MB/s (${times.length} iterations)`);
  return result;
}

async function main() {
  console.log('=== MongoDB Node.js driver (native) benchmarks ===');
  const results: BenchResult[] = [];

  const smallDoc = JSON.parse(readFileSync(join(DATA_DIR, 'small_doc.json'), 'utf-8'));
  const tweetDoc = JSON.parse(readFileSync(join(DATA_DIR, 'tweet.json'), 'utf-8'));
  const largeDoc = JSON.parse(readFileSync(join(DATA_DIR, 'large_doc.json'), 'utf-8'));
  const smallSize = Buffer.byteLength(JSON.stringify(smallDoc));
  const tweetSize = Buffer.byteLength(JSON.stringify(tweetDoc));
  const largeSize = Buffer.byteLength(JSON.stringify(largeDoc));

  // Run Command (batch 10,000 hello commands per iteration)
  results.push(await runBenchmark(
    'run_command', 'single_doc',
    async (_client) => {},
    async (_client) => {},
    async (client) => {
      for (let i = 0; i < 10_000; i++) {
        await client.db(DB_NAME).command({ hello: 1 });
      }
    },
    async (_client) => {},
    async (_client) => {},
    10_000 * 100, 10_000,
  ));

  // Find One by ID (batch 10,000 finds per iteration)
  results.push(await runBenchmark(
    'find_one_by_id', 'single_doc',
    async (client) => {
      const coll = client.db(DB_NAME).collection('bench_find');
      await coll.drop().catch(() => {});
      await coll.insertOne({ _id: new ObjectId('000000000000000000000001'), ...tweetDoc });
    },
    async (_client) => {},
    async (client) => {
      for (let i = 0; i < 10_000; i++) {
        await client.db(DB_NAME).collection('bench_find').findOne({ _id: new ObjectId('000000000000000000000001') });
      }
    },
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_find').drop().catch(() => {});
    },
    10_000 * tweetSize, 10_000,
  ));

  // InsertOne Small (batch 10,000 inserts per iteration)
  results.push(await runBenchmark(
    'insert_one_small', 'single_doc',
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_insert_small').drop().catch(() => {});
    },
    async (client) => {
      for (let i = 0; i < 10_000; i++) {
        await client.db(DB_NAME).collection('bench_insert_small').insertOne({ ...smallDoc, _id: new ObjectId() });
      }
    },
    async (_client) => {},
    async (_client) => {},
    10_000 * smallSize, 10_000,
  ));

  // InsertOne Large (batch 10 inserts per iteration)
  results.push(await runBenchmark(
    'insert_one_large', 'single_doc',
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_insert_large').drop().catch(() => {});
    },
    async (client) => {
      for (let i = 0; i < 10; i++) {
        await client.db(DB_NAME).collection('bench_insert_large').insertOne({ ...largeDoc, _id: new ObjectId() });
      }
    },
    async (_client) => {},
    async (_client) => {},
    10 * largeSize, 10,
  ));

  // Bulk Insert Small (10,000 docs per iteration)
  results.push(await runBenchmark(
    'bulk_insert_small', 'multi_doc',
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_bulk').drop().catch(() => {});
    },
    async (client) => {
      const docs = Array.from({ length: 10_000 }, () => ({ ...smallDoc, _id: new ObjectId() }));
      await client.db(DB_NAME).collection('bench_bulk').insertMany(docs);
    },
    async (_client) => {},
    async (_client) => {},
    smallSize * 10_000, 10_000,
  ));

  // Find Many (10,000 docs)
  results.push(await runBenchmark(
    'find_many', 'multi_doc',
    async (client) => {
      const coll = client.db(DB_NAME).collection('bench_find_many');
      await coll.drop().catch(() => {});
      const docs = Array.from({ length: 10_000 }, () => ({ ...smallDoc, _id: new ObjectId() }));
      await coll.insertMany(docs);
    },
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_find_many').find({}).toArray();
    },
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_find_many').drop().catch(() => {});
    },
    smallSize * 10_000, 10_000,
  ));

  // Bulk Insert Large (10 x ~2.75MB docs per iteration)
  results.push(await runBenchmark(
    'bulk_insert_large', 'multi_doc',
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_bulk_large').drop().catch(() => {});
    },
    async (client) => {
      const docs = Array.from({ length: 10 }, () => ({ ...largeDoc, _id: new ObjectId() }));
      await client.db(DB_NAME).collection('bench_bulk_large').insertMany(docs);
    },
    async (_client) => {},
    async (_client) => {},
    largeSize * 10, 10,
  ));

  // Find Many Large (10 x ~2.75MB docs)
  results.push(await runBenchmark(
    'find_many_large', 'multi_doc',
    async (client) => {
      const coll = client.db(DB_NAME).collection('bench_find_many_large');
      await coll.drop().catch(() => {});
      const docs = Array.from({ length: 10 }, () => ({ ...largeDoc, _id: new ObjectId() }));
      await coll.insertMany(docs);
    },
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_find_many_large').find({}).toArray();
    },
    async (_client) => {},
    async (client) => {
      await client.db(DB_NAME).collection('bench_find_many_large').drop().catch(() => {});
    },
    largeSize * 10, 10,
  ));

  // Save results
  const outputPath = join(RESULTS_DIR, 'typescript_native.json');
  writeFileSync(outputPath, JSON.stringify(results, null, 2));
  console.log(`\nResults saved to ${outputPath}`);
}

main().catch((err) => { console.error(err); process.exit(1); });

File: benchmarks/drivers/python/bench_native.py
"""Benchmark pymongo (native driver) for comparison against MongoCore."""

import json
import os
import platform
import sys
import time
import statistics
from datetime import datetime, timezone
from pathlib import Path

from pymongo import MongoClient
from bson import ObjectId

# Load config
CONFIG = json.loads((Path(__file__).parent.parent / "common.json").read_text())
DATA_DIR = Path(__file__).parent.parent.parent / "data"
RESULTS_DIR = Path(__file__).parent.parent.parent / "results"
RESULTS_DIR.mkdir(exist_ok=True)

QUICK_MODE = "--quick" in sys.argv

WARMUP = 0 if QUICK_MODE else CONFIG["warmup_iterations"]["python"]
MIN_TIME = 0 if QUICK_MODE else CONFIG["min_time_secs"]
MAX_ITERS = 1 if QUICK_MODE else CONFIG["max_iterations"]
MAX_TIME = 5 if QUICK_MODE else CONFIG["max_time_secs"]
DB_NAME = CONFIG["database"]


def get_system_info():
    return {
        "os": platform.system().lower(),
        "arch": platform.machine(),
        "cpus": os.cpu_count(),
        "ram_gb": round(os.sysconf("SC_PAGE_SIZE") * os.sysconf("SC_PHYS_PAGES") / (1024**3), 1) if hasattr(os, "sysconf") else None,
        "mongocore_version": "native",
        "driver": "pymongo",
    }


def run_benchmark(name, category, setup_fn, before_task_fn, task_fn, after_task_fn, teardown_fn, dataset_size_bytes, batch_size=1):
    """Run a benchmark following MongoDB spec methodology."""
    client = MongoClient(CONFIG["mongodb_uri"], w=1)
    db = client[DB_NAME]

    setup_fn(db)

    # Warmup
    for _ in range(WARMUP):
        before_task_fn(db)
        task_fn(db)
        after_task_fn(db)

    # Timed iterations
    times = []
    total_time = 0.0
    iteration = 0

    while total_time < MIN_TIME or iteration < 5:
        if iteration >= MAX_ITERS or total_time >= MAX_TIME:
            break

        before_task_fn(db)

        start = time.perf_counter()
        task_fn(db)
        elapsed = time.perf_counter() - start

        after_task_fn(db)

        times.append(elapsed)
        total_time += elapsed
        iteration += 1

    teardown_fn(db)
    client.close()

    # Calculate metrics
    times.sort()
    median = statistics.median(times)
    ops_per_sec = batch_size / median
    mb_per_sec = dataset_size_bytes / median / 1_000_000

    def percentile(data, pct):
        import math
        idx = max(0, math.ceil(len(data) * pct / 100) - 1)
        return data[min(idx, len(data) - 1)]

    result = {
        "benchmark": name,
        "category": category,
        "driver": "pymongo",
        "dataset_size_bytes": dataset_size_bytes,
        "batch_size": batch_size,
        "iterations": len(times),
        "total_time_secs": round(total_time, 3),
        "ops_per_sec": round(ops_per_sec, 1),
        "mb_per_sec": round(mb_per_sec, 3),
        "percentiles": {
            "p10": round(percentile(times, 10), 6),
            "p25": round(percentile(times, 25), 6),
            "p50": round(median, 6),
            "p75": round(percentile(times, 75), 6),
            "p90": round(percentile(times, 90), 6),
            "p95": round(percentile(times, 95), 6),
            "p99": round(percentile(times, 99), 6),
        },
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "system": get_system_info(),
    }

    print(f"  {name}: {ops_per_sec:.0f} ops/s, {mb_per_sec:.2f} MB/s ({len(times)} iterations)")
    return result


def main():
    print("=== pymongo (native) benchmarks ===")
    results = []

    small_doc = json.loads((DATA_DIR / "small_doc.json").read_text())
    tweet_doc = json.loads((DATA_DIR / "tweet.json").read_text())
    large_doc = json.loads((DATA_DIR / "large_doc.json").read_text())
    small_size = len(json.dumps(small_doc).encode())
    tweet_size = len(json.dumps(tweet_doc).encode())
    large_size = len(json.dumps(large_doc).encode())

    # Run Command (batch 10,000 hello commands per iteration)
    def task_run_command(db):
        for _ in range(10_000):
            db.command("hello")

    results.append(run_benchmark(
        "run_command", "single_doc",
        setup_fn=lambda db: None,
        before_task_fn=lambda db: None,
        task_fn=task_run_command,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: None,
        dataset_size_bytes=10_000 * 100, batch_size=10_000,
    ))

    # Find One by ID (batch 10,000 finds per iteration)
    def setup_find(db):
        db.drop_collection("bench_find")
        coll = db["bench_find"]
        doc = {k: v for k, v in tweet_doc.items() if k != "_id"}
        doc["_id"] = ObjectId("000000000000000000000001")
        coll.insert_one(doc)
        coll.create_index("_id")

    def task_find_one(db):
        for _ in range(10_000):
            db["bench_find"].find_one({"_id": ObjectId("000000000000000000000001")})

    results.append(run_benchmark(
        "find_one_by_id", "single_doc",
        setup_fn=setup_find,
        before_task_fn=lambda db: None,
        task_fn=task_find_one,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_find"),
        dataset_size_bytes=10_000 * tweet_size, batch_size=10_000,
    ))

    # InsertOne Small (batch 10,000 inserts per iteration)
    def task_insert_small(db):
        for _ in range(10_000):
            db["bench_insert_small"].insert_one({**small_doc, "_id": ObjectId()})

    results.append(run_benchmark(
        "insert_one_small", "single_doc",
        setup_fn=lambda db: None,
        before_task_fn=lambda db: db.drop_collection("bench_insert_small"),
        task_fn=task_insert_small,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_insert_small"),
        dataset_size_bytes=10_000 * small_size, batch_size=10_000,
    ))

    # InsertOne Large (batch 10 inserts per iteration, large docs ~2.75MB each)
    def task_insert_large(db):
        for _ in range(10):
            db["bench_insert_large"].insert_one({**large_doc, "_id": ObjectId()})

    results.append(run_benchmark(
        "insert_one_large", "single_doc",
        setup_fn=lambda db: None,
        before_task_fn=lambda db: db.drop_collection("bench_insert_large"),
        task_fn=task_insert_large,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_insert_large"),
        dataset_size_bytes=10 * large_size, batch_size=10,
    ))

    # Bulk Insert Small (10,000 docs per iteration)
    def bulk_insert_task(db):
        docs = [{**small_doc, "_id": ObjectId()} for _ in range(10_000)]
        db["bench_bulk"].insert_many(docs)

    results.append(run_benchmark(
        "bulk_insert_small", "multi_doc",
        setup_fn=lambda db: None,
        before_task_fn=lambda db: db.drop_collection("bench_bulk"),
        task_fn=bulk_insert_task,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_bulk"),
        dataset_size_bytes=small_size * 10_000, batch_size=10_000,
    ))

    # Bulk Insert Large (10 x 2.75MB docs per iteration)
    def bulk_insert_large_task(db):
        docs = [{**large_doc, "_id": ObjectId()} for _ in range(10)]
        db["bench_bulk_large"].insert_many(docs)

    results.append(run_benchmark(
        "bulk_insert_large", "multi_doc",
        setup_fn=lambda db: None,
        before_task_fn=lambda db: db.drop_collection("bench_bulk_large"),
        task_fn=bulk_insert_large_task,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_bulk_large"),
        dataset_size_bytes=large_size * 10, batch_size=10,
    ))

    # Find Many (10,000 small docs)
    def setup_find_many(db):
        db.drop_collection("bench_find_many")
        docs = [{**small_doc, "_id": ObjectId()} for _ in range(10_000)]
        db["bench_find_many"].insert_many(docs)

    def find_many_task(db):
        list(db["bench_find_many"].find({}))

    results.append(run_benchmark(
        "find_many", "multi_doc",
        setup_fn=setup_find_many,
        before_task_fn=lambda db: None,
        task_fn=find_many_task,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_find_many"),
        dataset_size_bytes=small_size * 10_000, batch_size=10_000,
    ))

    # Find Many Large (100 large docs, emptying cursor)
    def setup_find_many_large(db):
        db.drop_collection("bench_find_many_large")
        docs = [{**large_doc, "_id": ObjectId()} for _ in range(10)]
        db["bench_find_many_large"].insert_many(docs)

    def find_many_large_task(db):
        list(db["bench_find_many_large"].find({}))

    results.append(run_benchmark(
        "find_many_large", "multi_doc",
        setup_fn=setup_find_many_large,
        before_task_fn=lambda db: None,
        task_fn=find_many_large_task,
        after_task_fn=lambda db: None,
        teardown_fn=lambda db: db.drop_collection("bench_find_many_large"),
        dataset_size_bytes=large_size * 10, batch_size=10,
    ))

    # Save results
    output_path = RESULTS_DIR / "python_native.json"
    with open(output_path, "w") as f:
        json.dump(results, f, indent=2)
    print(f"\nResults saved to {output_path}")


if __name__ == "__main__":
    main()
